### PR TITLE
[spirv] Add CompositeInsertOp operation.

### PR DESCRIPTION
**Pull request by denis0x0D**
A CompositeInsertOp operation make a copy of a composite object,
while modifying one part of it.

@antiagainst @MaheshRavishankar can you please take a look, thanks!

### CODE REVIEW

File: include/mlir/Dialect/SPIRV/SPIRVOps.td [hunk 1]
```diff
@@ -153,6 +153,52 @@ def SPV_CompositeExtractOp : SPV_Op<"CompositeExtract", [NoSideEffect]> {
 
 // -----
 
+def SPV_CompositeInsertOp : SPV_Op<"CompositeInsert", [NoSideEffect]> {
```
Comment (antiagainst) [inline include/mlir/Dialect/SPIRV/SPIRVOps.td]: Could you create a new file, `SPIRVCompositeOps.td`, and put the definition there to follow spec organization?

File: lib/Dialect/SPIRV/SPIRVOps.cpp [hunk 2]
```diff
@@ -1058,6 +1058,83 @@ OpFoldResult spirv::CompositeExtractOp::fold(ArrayRef<Attribute> operands) {
...
+
+static ParseResult parseCompositeInsertOp(OpAsmParser &parser,
+                                          OperationState &state) {
+  OpAsmParser::OperandType objectInfo, compositeInfo;
+  Type objectType, compositeType;
+  Attribute indicesAttr;
+  auto loc = parser.getCurrentLocation();
+
+  if (parser.parseOperand(objectInfo) || parser.parseComma() ||
+      parser.parseOperand(compositeInfo) ||
```
Comment (antiagainst) [inline lib/Dialect/SPIRV/SPIRVOps.cpp]: `parseOperandList(..., 2)`?

File: lib/Dialect/SPIRV/SPIRVOps.cpp [hunk 3]
```diff
@@ -1058,6 +1058,83 @@ OpFoldResult spirv::CompositeExtractOp::fold(ArrayRef<Attribute> operands) {
...
+      parser.parseAttribute(indicesAttr, kIndicesAttrName, state.attributes) ||
+      parser.parseColonType(objectType) ||
+      parser.parseKeywordType("into", compositeType) ||
+      parser.resolveOperands({objectInfo, compositeInfo},
+                             {objectType, compositeType}, loc,
+                             state.operands)) {
+    return failure();
+  }
+
+  state.addTypes(compositeType);
```
Comment (antiagainst) [inline lib/Dialect/SPIRV/SPIRVOps.cpp]: return parser.addTypeToList(compositeType, state.types);

File: lib/Dialect/SPIRV/SPIRVOps.cpp [hunk 4]
```diff
@@ -1058,6 +1058,83 @@ OpFoldResult spirv::CompositeExtractOp::fold(ArrayRef<Attribute> operands) {
...
+}
+
+static LogicalResult verify(spirv::CompositeInsertOp compositeInsertOp) {
+  auto indicesArrayAttr = compositeInsertOp.indices().dyn_cast<ArrayAttr>();
+  if (!indicesArrayAttr.size()) {
+    return compositeInsertOp.emitOpError(
+        "expected at least one index for spv.CompositeInsert");
+  }
+
+  // TODO: `spv.CompositeInsert` and `spv.CompositeExtrat` share the same logic
```
Comment (antiagainst) [inline lib/Dialect/SPIRV/SPIRVOps.cpp]: Yes please! :)

File: test/Dialect/SPIRV/ops.mlir [hunk 5]
```diff
@@ -420,6 +420,50 @@ func @composite_extract_result_type_mismatch(%arg0: !spv.array<4xf32>) -> i32 {
 
 // -----
 
+//===----------------------------------------------------------------------===//
```
Comment (antiagainst) [inline test/Dialect/SPIRV/ops.mlir]: Could you create a new file `composite-ops.mlir` to host these tests?

File: lib/Dialect/SPIRV/SPIRVOps.cpp [hunk 6]
```diff
@@ -1058,6 +1058,83 @@ OpFoldResult spirv::CompositeExtractOp::fold(ArrayRef<Attribute> operands) {
...
+    }
+  }
+
+  if (objectType != compositeInsertOp.object()->getType()) {
+    return compositeInsertOp.emitOpError("invalid object type: expected ")
+           << objectType << " but provided "
+           << compositeInsertOp.object()->getType();
+  }
+
+  if (compositeInsertOp.composite()->getType() != compositeInsertOp.getType()) {
```
Comment (antiagainst) [inline lib/Dialect/SPIRV/SPIRVOps.cpp]: Test for this?

File: lib/Dialect/SPIRV/SPIRVOps.cpp [hunk 7]
```diff
@@ -1058,6 +1058,83 @@ OpFoldResult spirv::CompositeExtractOp::fold(ArrayRef<Attribute> operands) {
...
+    return compositeInsertOp.emitOpError("invalid result type: expected ")
```
Comment (antiagainst) [inline lib/Dialect/SPIRV/SPIRVOps.cpp]: "result type should be the same as the composite operand type, but found " << ... << " vs " << ... ?

File: lib/Dialect/SPIRV/SPIRVOps.cpp [hunk 8]
```diff
@@ -1058,6 +1058,83 @@ OpFoldResult spirv::CompositeExtractOp::fold(ArrayRef<Attribute> operands) {
...
+      objectType = cType.getElementType(index);
+    } else {
+      return compositeInsertOp.emitError(
+                 "cannot extract from non-composite type ")
+             << objectType << " with index " << index;
+    }
+  }
+
+  if (objectType != compositeInsertOp.object()->getType()) {
+    return compositeInsertOp.emitOpError("invalid object type: expected ")
```
Comment (antiagainst) [inline lib/Dialect/SPIRV/SPIRVOps.cpp]: "object operand type should be " << objectType << " but found " << ... ?

**Comment by denis0x0D** [comment]
@antiagainst thanks a lot for review! Updated.

File: lib/Dialect/SPIRV/SPIRVOps.cpp [hunk 9]
```diff
@@ -953,6 +953,31 @@ static LogicalResult verify(spirv::BranchConditionalOp branchOp) {
 // spv.CompositeExtractOp
 //===----------------------------------------------------------------------===//
 
+static Type getElementType(Type type, ArrayAttr indices, Location loc) {
```
Comment (antiagainst) [inline lib/Dialect/SPIRV/SPIRVOps.cpp]: Could you move this to the top of this file and put it under the "Common utility functions" section?

Comment (antiagainst) [inline lib/Dialect/SPIRV/SPIRVOps.cpp]: Missing doc for this function.

"Walks walk the given `type` hierarchy with the given `indices`, potentially down to component granularity, to select an element type. Returns null type and emits errors with the given `loc` on failure."

Comment (denis0x0D) [reply lib/Dialect/SPIRV/SPIRVOps.cpp]: Thansk, fixed!

Comment (denis0x0D) [reply lib/Dialect/SPIRV/SPIRVOps.cpp]: Thanks!

**Comment by denis0x0D** [comment]
Seems like tests failed because of this error:
/tmpfs/src/git/llvm/llvm/projects/mlir/test/lib/IR/TestSymbolUses.cpp:18:25: fatal error: TestDialect.h: No such file or directory
compilation terminated.
projects/mlir/test/lib/IR/CMakeFiles/MLIRTestIR.dir/build.make:82: recipe for target 'projects/mlir/test/lib/IR/CMakeFiles/MLIRTestIR.dir/TestSymbolUses.cpp.o' failed

**Comment by antiagainst** [comment]
There is also a conflict. Could you rebase against master?

**Pull request closed without merging**